Port TFruityPlugHost PlugMsg methods (#51)

dmitryrec opened the pull request:
Closes #51 
Closes #85

dmitryrec committed:
Init PR

dmitryrec committed:
resolve conflict

dmitryrec committed:
Merge branch '51-port-plugmsg-methods' of https://github.com/tonikasoft/fpsdk into 51-port-plugmsg-methods

dmitryrec committed:
Update install script for Windows
- fix width line in wrapper.cpp

dmitryrec committed:
Update install script for Windows
- fix width line in wrapper.cpp

dmitryrec committed:
Update install script for Windows
- fix width line in wrapper.cpp

dmitryrec committed:
Merge branch '51-port-plugmsg-methods' of https://github.com/tonikasoft/fpsdk into 51-port-plugmsg-methods

dmitryrec committed:
Implement loop_out
- fix bat file

dmitryrec committed:
Implement loop_out
- fix bat file

dmitryrec committed:
Merge branch 'master' of https://github.com/tonikasoft/fpsdk

dmitryrec committed:
Resolve conflicts

dmitryrec committed:
Update loop_out example
- find bat file bug

dmitryrec committed:
Add loop_kill
- add doc to loop_out

dmitryrec committed:
Fix loop_out doc

dmitryrec committed:
Update and test bat file

ales-tsurko commented on examples/simple.rs:
```diff
@@ -199,17 +199,24 @@ impl Plugin for Simple {
                 // self.host.on_message(self.tag, message::ActivateMidi);
+                self.host.loop_out(
+                    self.tag,
+                    ValuePtr::from_raw_ptr(String::from("W").as_raw_ptr()),
```
Я думал, ты что-то более информативное потом в эту строку напишешь. Сделай так:

```Rust
ValuePtr::from_raw_ptr(format!("{:?}", message).as_raw_ptr()),
```

ales-tsurko replied:
Потом ниже имплементация `loop_in` у тебя как была, так и осталась. Суть же в том, что ты передаешь строку с помощью `loop_out` и там ее выводишь в лог, например.

Еще нужно добавить в `InfoBuilder` выше `loop_out` все-таки. Потому что как-то нелогично получается, что мы показываем, как с ним работать, говорим, что не работает, но при этом сами ошибку в работе допускаем.

ales-tsurko commented on install.win.bat:
```diff
@@ -8,20 +8,36 @@ set dest_name=%2
 set plugins_dir="C:\Program Files\Image-Line\FL Studio 20\Plugins\Fruity"
 set args_count=0
 set flag=false
```
Теперь непонятно, что это за `flag`. Придумай какой-нибудь понятное название переменной. Я уже смутно помню, что это такое, а было бы нормальное имя — было бы понятно. Помню, что вроде для проверки наличия опционального параметра.

ales-tsurko commented on src/host.rs:
```diff
@@ -159,6 +159,20 @@ impl Host {
     }
+
+    /// Ask for a message to be dispatched to itself when the current mixing tick will be played (to
+    /// synchronize stuff) (see [`loop_in`](../plugin/trait.Plugin.html#method.loop_in)). 
```
Блин, я тебе целый день говорю про это. Что про коммиты, что здесь:

просто `loop_in` — это вообще ни о чем. Особенно в контексте доков. Ты так написал, что это получается тупо функция. А речь идет о методе `Plugin::loop_in`.

ales-tsurko commented on src/host.rs:
```diff
@@ -159,6 +159,20 @@ impl Host {
         Arc::clone(&self.out_voicer)
     }
+
+    /// Ask for a message to be dispatched to itself when the current mixing tick will be played (to
```
Так же как с `Plugin::loop_in` укажи, что он возможно не работает. Сверху жирным капсом.

ales-tsurko commented on src/host.rs:
```diff
@@ -220,6 +234,8 @@ impl SendVoiceHandler for Voicer {
 extern "C" {
     fn host_release_voice(host: *mut c_void, tag: intptr_t);
     fn host_kill_voice(host: *mut c_void, tag: intptr_t);
+    fn host_loop_out(host: *mut c_void, tag: intptr_t, message: intptr_t);
+    fn host_loop_kill(host: *mut c_void, tag: intptr_t, message: intptr_t);
```
Ну выше ж есть очевидный блок, где место этим функциям, зачем их пихать к войсам? Какое эти функции имеют к ним отношение?

Там и внутри `Host` логичнее компановать их с методами MIDI out, например, а не обволакивать методы для войсов методами синхронизаций.

ales-tsurko commented on install.win.bat:
```diff
@@ -1,5 +1,5 @@
 :: Usage:
-:: install.win.bat name destination_name [plugins_dir]
+:: install.win.bat name destination_name type_of_plugin [plugins_dir]
```
👍

ales-tsurko commented on src/cxx/wrapper.h:
```diff
@@ -128,6 +128,8 @@ extern "C" void host_midi_out_del(void *host, TPluginTag tag, unsigned char
 
 extern "C" void host_release_voice(void *host, intptr_t tag);
 extern "C" void host_kill_voice(void *host, intptr_t tag);
+extern "C" void host_loop_out(void *host, TPluginTag tag, intptr_t msg);
```
И тут в войсы засунул.

ales-tsurko commented on src/cxx/wrapper.cpp:
```diff
@@ -289,6 +289,14 @@ void host_kill_voice(void *host, intptr_t tag) {
     ((TFruityPlugHost *)host)->Voice_Kill(tag, true);
 }
 
+void host_loop_out(void *host, TPluginTag tag, intptr_t msg) {
+    ((TFruityPlugHost *)host)->PlugMsg_Delayed(tag, msg);
+}
+
+void host_loop_kill(void *host, TPluginTag tag, intptr_t msg) {
+    ((TFruityPlugHost *)host)->PlugMsg_Kill(tag, msg);
+}
```
И тут в войсах...

ales-tsurko commented on src/host.rs:
```diff
@@ -159,6 +159,20 @@ impl Host {
+        unsafe { host_loop_out(*self.host_ptr.get_mut(), tag.0, message.0) };
+    }
+
+    ///Remove a buffered message, so that it will never be dispatched.
```
Пробел перед "Remove".

Вообще лучше написать:

Remove the buffered message scheduled by `Host::loop_out`, so that it will never be dispatched.

dmitryrec committed:
Rename flag for correct args count in bat file
- fix doc for Plugin::loop_out
- fix implementation for Plugin::loop_in
- add InfoBuilder:: loop_out

ales-tsurko commented on install.win.bat:
```diff
@@ -1,27 +1,43 @@
-    echo "Usage: install.win.bat name destination_name [plugins_dir]"
+if %args_count% lss 3 set correct_args_count=true
+if %args_count% gtr 4 set correct_args_count=true
+if "%correct_args_count%"=="true" (
```
То есть если количество аргументов правильное, выводим ошибку? 😄 

Так понятнее. Только получается, что `incorrect_` должно быть.

ales-tsurko commented on src/cxx/wrapper.h:
```diff
@@ -125,9 +125,11 @@ extern "C" void host_midi_out(void *host, TPluginTag tag, unsigned char status,
                               unsigned char port);
 extern "C" void host_midi_out_del(void *host, TPluginTag tag, unsigned char
         status, unsigned char data1, unsigned char data2, unsigned char port);
+extern "C" void host_kill_voice(void *host, intptr_t tag);
```
Зачем ты `kill_voice` сюда перекинул? Еще больше все запутал.

ales-tsurko commented on src/cxx/wrapper.h:
```diff
@@ -125,9 +125,11 @@ extern "C" void host_midi_out(void *host, TPluginTag tag, unsigned char status,
 
 extern "C" void host_release_voice(void *host, intptr_t tag);
-extern "C" void host_kill_voice(void *host, intptr_t tag);
+extern "C" void host_loop_kill(void *host, TPluginTag tag, intptr_t msg);
```
Опять в войсах.

ales-tsurko commented on src/host.rs:
```diff
@@ -159,6 +159,24 @@ impl Host {
+    ///
+    /// The message is guaranteed to be dispatched, however it could be sent immediately if it
+    /// couldn't be buffered (it's only buffered when playing).
+    pub fn loop_out(&mut self, tag: plugin::Tag, message: ValuePtr) {
+        unsafe { host_loop_out(*self.host_ptr.get_mut(), tag.0, message.0) };
+    }
+
+    /// Remove the buffered message scheduled by Host::loop_out.
+    pub fn loop_kill(&mut self, tag: plugin::Tag, message: ValuePtr) {
+        unsafe { host_loop_kill(*self.host_ptr.get_mut(), tag.0, message.0) };
+    }
```
Еще раз говорю: перекинь эти методы так, чтобы они были перед войсами и после MIDI out (где им более подходящее место и методы, которые у нас на треде MIDI не будут обволакивать методы, относящиеся к войсам). Плюс таким образом у тебя код будет соответствовать стайл гайду. Иначе получается, что FFI функции, которые вызываются у тебя в этих методах — не по гайду.

dmitryrec committed:
Rename flag in bat file
- replace host_loop_kill and host_loop_out in correct places
- add link in Host::loop_out doc
